### ISSUE TITLE
Model Seems Severely Overfitted - Poor Real-World Performance

### ISSUE THREAD

**Issue by sabman** [Roysubh/Coconut-Tree-Detection-from-High-Resolution-Imagery-Using-YOLO-in-Python#1]
#  Possible Over-fitting - Poor Real-World Performance

Thanks for sharing this model! We just tried this with some real-world data and discovered the model is severely overfitted, showing poor inference results and numerous false positives.

### Training Metrics Analysis (Final Epoch 2000)
| Metric | Training | Validation | Gap |
|--------|----------|------------|-----|
| Box Loss | 1.037 | 1.480 | **+43%** |
| DFL Loss | 1.124 | 1.404 | **+25%** |
| mAP50 | 85.0% | - | - |

### Real-World Testing Results
**Test image attached** - 1952x2033 GeoTIFF from Southeast Asia

```bash
# Results on our test data
Confidence 0.05: 11 trees detected (max conf: 38.5%)
Confidence 0.01: 90 trees detected (79 false positives!)
```

**Issues We Observed:**
- 50+ detections clustered at image edges (x=1952)
- Systematic boundary artifacts at `[1914.2, *, 1952.0, *]`
- Confidence scores as low as 0.01-0.02
- Model learned training image boundaries instead of coconut features

## Root Cause Analysis
1. **Excessive epochs**: 2000 vs recommended 200-500 for YOLO
2. **No early stopping**: Validation loss diverged significantly from training
3. **Single scale training**: Only 256x256 patches
4. **Insufficient regularization**: Model memorized training patterns

## Impact on Our Use Case
- Low confidence scores (0.05-0.38) make production deployment difficult
- Massive false positive rate at reasonable thresholds
- Edge artifacts make model unusable for full-scene inference
- ONNX deployment affected by base model quality

## Suggested Improvements
1. **Retrain with early stopping** (max 300 epochs, patience=50)
2. **Multi-scale training** (256, 512, 640px)
3. **Proper validation monitoring** to prevent overfitting
4. **Enhanced augmentation** and regularization
5. **Stop training when val_loss plateaus**

## Would Love to See
- [ ] Val loss within 10% of training loss
- [ ] Confidence scores >30% for real detections  
- [ ] <5% false positives at conf>0.1
- [ ] No edge artifacts on diverse test imagery

Thanks again for the great work - happy to help test any retrained versions!

---
*Test image attached: 1952x2033 GeoTIFF showing the overfitting artifacts we discovered*

https://oin-hotosm-temp.s3.us-east-1.amazonaws.com/65c6eb328931500001717ddc/0/65c6eb328931500001717ddd.tif

**Comment by Roysubh** [comment]
Hi [sabman](https://github.com/sabman)

Thanks for testing the model and sharing your observations. I’d like to address your points step by step:

1. Resolution difference:

- Your test image is 50 cm, whereas the model was trained on 9 cm high-resolution imagery.
- YOLO models are very sensitive to spatial resolution. Applying a 9 cm-trained model on 50 cm imagery naturally reduces confidence scores and increases false positives.

2. Epoch count does not automatically indicate overfitting:

- Using 2000 epochs does not automatically mean the model is overfitted. Overfitting occurs when a model performs well on training data but poorly on unseen data of the same distribution.
- I tested the model on unseen 9 cm areas (not used in training), and it still performed well, which shows the model generalizes correctly.

3. Early stopping argument:

- Early stopping is a regularization technique to prevent overfitting by monitoring validation loss.
- Not using early stopping does not automatically cause overfitting. The real test is how the model performs on new, unseen data of the same type.
- As noted, the model still performs well on unseen 9 cm imagery, confirming it is not overfitted.

4. Validation metrics:

- Precision = 0.886 → most predicted trees are correct.
- Recall = 0.802 → majority of actual trees detected.
- mAP@0.5 = 0.850 → strong bounding box alignment.
- These metrics demonstrate high-quality detections with minimal false positives.

5. Edge artifacts / low confidence detections:

- The systematic boundary detections and low confidence scores on your 50 cm image are likely caused by patching/tiling differences and the resolution mismatch.
- These issues are not indicative of overfitting; they are domain and scale effects.

6. Comparison of results:
I have attached your 50 cm data vs my 9 cm data, both viewed in ArcMap at the same scale (1:300).
This clearly shows the difference in detection quality and confirms the model generalizes at its intended resolution.

<img width="720" height="869" alt="Image" src="https://github.com/user-attachments/assets/d7745582-62d7-4b25-aaaf-7462024c3fc6" />

✅ Conclusion:
The model is not overfitted.
The low-confidence detections you observed are due to applying a 9 cm-trained model to 50 cm imagery, not memorization.
For best results, the model should be applied to imagery of similar resolution (~9 cm).

Thanks,
Subham Roy

**Comment by Roysubh** [comment]
Hi [sabman](https://github.com/sabman),

Thank you very much for testing the model and sharing your observations. After reviewing, it appears that the low-confidence detections are due to the difference in resolution (50 cm test image vs 9 cm training data), rather than overfitting. The model performs well on unseen 9 cm imagery, as shown in the attached comparison.

I’m closing this issue now, but please feel free to reopen or comment if there are any further concerns.

<img width="720" height="869" alt="Image" src="https://github.com/user-attachments/assets/f8b4b071-fbc7-46aa-8346-454ed53df893" />

Best regards,
Subham Roy
